File: C++/similar-rgb-color.cpp
// Time:  O(1)
// Space: O(1)

class Solution {
public:
    string similarRGB(string color) {
        string result = "#";
        result += round(color.substr(1, 2));
        result += round(color.substr(3, 2));
        result += round(color.substr(5, 2));
        return result;
    }

private:
    string round(const string& color) {
        std::stringstream in(color);
        int decimal = 0;
        in >> std::hex >> decimal;
        decimal = decimal / 17 + (decimal % 17 > 8 ? 1 : 0);
        std::stringstream out;
        out << std::setfill('0') << std::setw(2) << std::hex <<  17 * q;
        return out.str();
    }
};

File: Python/similar-rgb-color.py
# Time:  O(1)
# Space: O(1)

class Solution(object):
    def similarRGB(self, color):
        """
        :type color: str
        :rtype: str
        """
        def rounding(color):
            q, r = divmod(int(color, 16), 17)
            if r > 8: q += 1
            return '{:02x}'.format(17*q)

        return '#' + \
                rounding(color[1:3]) + \
                rounding(color[3:5]) + \
                rounding(color[5:7])

